File: lib/server/locks.ts
import * as db from "./db";
import { normalizePrefix } from "./r2";
import { ApiError, unlockedFolders } from "./security";

export function lockingAncestor(path: string, locked?: string[]): string | null {
  const list = locked ?? [];
  const matches = list.filter((f) => path === f || path.startsWith(f));
  if (!matches.length) return null;
  return matches.reduce((a, b) => (a.length >= b.length ? a : b));
}

export async function ensureUnlocked(
  path: string,
  folderToken: string | null,
  userId: string,
): Promise<void> {
  const locked = await db.listLockedFolders(userId);
  const gate = lockingAncestor(path, locked);
  if (!gate) return;
  const unlocked = await unlockedFolders(folderToken, userId);
  if (!unlocked.has(gate)) {
    throw new ApiError(423, { message: "This folder is password protected.", folder: gate });
  }
}

export function folderTokenFromRequest(req: Request): string | null {
  return req.headers.get("x-folder-token");
}

export async function isLockedPrefix(prefix: string, userId: string): Promise<boolean> {
  const locked = await db.listLockedFolders(userId);
  return locked.includes(normalizePrefix(prefix));
}

File: backend/app/locks.py
"""Folder password-lock enforcement helpers."""

from __future__ import annotations

from fastapi import Header, HTTPException, status

from . import db
from .r2 import normalize_prefix
from .security import unlocked_folders


def locking_ancestor(path: str, locked: list[str] | None = None) -> str | None:
    """Return the nearest locked folder that gates ``path`` (a key or prefix).

    A folder ``locked`` gates ``path`` when ``path`` is the folder itself or
    lives anywhere beneath it.
    """
    locked = locked if locked is not None else []
    matches = [f for f in locked if path == f or path.startswith(f)]
    if not matches:
        return None
    return max(matches, key=len)


def ensure_unlocked(path: str, folder_token: str | None, user_id: str) -> None:
    """Raise 423 (Locked) when ``path`` is gated by a folder password the
    caller has not unlocked."""
    locked = db.list_locked_folders(user_id)
    gate = locking_ancestor(path, locked)
    if gate is None:
        return
    if gate not in unlocked_folders(folder_token, user_id):
        raise HTTPException(
            status_code=status.HTTP_423_LOCKED,
            detail={"message": "This folder is password protected.", "folder": gate},
        )


def folder_token_header(
    x_folder_token: str | None = Header(default=None, alias="X-Folder-Token"),
) -> str | None:
    return x_folder_token


def is_locked_prefix(prefix: str, user_id: str) -> bool:
    return normalize_prefix(prefix) in set(db.list_locked_folders(user_id))
